File: Design-Patterns-Using-Ruby/Basics OOPS/Inheritance.rb
class Person
    attr_reader :name, :age, :gender

    def initialize(name, age, gender)
        @name = name
        @age = age
        @gender = gender
    end

    def to_s
        "Person: #{@name} is a #{@gender} and #{@age} years old."
    end
end

class Citizen < Person
    attr_reader :voter_id

    def initialize(name, age, voter_id, gender)
        super(name, age, gender)
        @voter_id = voter_id
    end

    def to_s
        "Citizen: #{@name} is a #{@gender} and #{@age} years old with voter id #{@voter_id}."
    end
end

# Client code
p = Person.new("John", 32, "Male")
puts p.to_s

c = Citizen.new("Smith", 31, 1234, "Male")
puts c.to_s

File: Design-Patterns-Using-Kotlin/Basics OOPS/Inheritance.kt
open class Person(protected var name: String, protected var age: Int, protected var gender: String) {
    override fun toString(): String {
        return String.format("Person: %s is a %s and %s years old.", name, gender, age)
    }
}

class Citizen(name: String, age: Int, gender: String, var voterId: Int) : Person(name, age, gender) {
    override fun toString(): String {
        return String.format("Citizen: %s is a %s and %s years old with voter id %s.", name, gender, age, voterId)
    }
}

fun main() {
    val p = Person("John", 32, "Male")
    println(p)

    val c = Citizen("Smith", 31, "Male", 1234)
    println(c)
}

/*
Person: John is a Male and 32 years old.
Citizen: Smith is a Male and 31 years old with voter id 1234.
*/